beacon-chain/cache/sync_subnet_ids.go

```diff
@@ -0,0 +1,120 @@
+package cache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/rand"
+	"github.com/prysmaticlabs/prysm/shared/sliceutil"
+)
+
+type syncSubnetIDs struct {
+	sCommittee    *cache.Cache
+	sCommiteeLock sync.RWMutex
+}
+
+// SyncSubnetIDs for sync committee participant.
+var SyncSubnetIDs = newSyncSubnetIDs()
+
+func newSyncSubnetIDs() *syncSubnetIDs {
+	epochDuration := time.Duration(params.BeaconConfig().SlotsPerEpoch.Mul(params.BeaconConfig().SecondsPerSlot))
+	subLength := epochDuration * time.Duration(params.BeaconConfig().EpochsPerSyncCommitteePeriod)
+	persistentCache := cache.New(subLength*time.Second, epochDuration*time.Second)
+	return &syncSubnetIDs{sCommittee: persistentCache}
+}
+
+// GetSyncCommitteeSubnets retrieves the sync committee subnet and expiration time of that validator's
+// subscription.
+func (s *syncSubnetIDs) GetSyncCommitteeSubnets(pubkey []byte, epoch types.Epoch) ([]uint64, types.Epoch, bool, time.Time) {
+	s.sCommiteeLock.RLock()
+	defer s.sCommiteeLock.RUnlock()
+
+	id, duration, ok := s.sCommittee.GetWithExpiration(keyBuilder(pubkey, epoch))
+	if !ok {
+		return []uint64{}, 0, ok, time.Time{}
+	}
+	// Retrieve the slot from the cache.
+	idxs, ok := id.([]uint64)
+	if !ok {
+		return []uint64{}, 0, ok, time.Time{}
+	}
+	// If no committees are saved, we return
+	// nothing.
+	if len(idxs) <= 1 {
+		return []uint64{}, 0, ok, time.Time{}
+	}
+	return idxs[1:], types.Epoch(idxs[0]), ok, duration
```

Comment:
Why does this skip the first `idxs`?

Reply:
Looks like idxs[0] is used as metadata. It's stored for `joinEpoch`, where it is defined as `select a random number of epochs before the end of the current sync committee period between 1 and SYNC_COMMITTEE_SUBNET_COUNT, inclusive`

Reply:
After reading the whole file then reading this again, I now know why. My suggestion is to add commentary here so the reader doesn't have to read the whole file to understand this